buildkite/benchmark/run.py

```diff
@@ -267,17 +351,141 @@ def execute_command(self, command, path=".", exit_on_failure=True, log_stdout=Tr
         logging.info(f"Done executing -> {command}")
         return return_code, stderr
 
+
+class BenchmarkGroupsRunner(ABC):
+    "Abstract class for runners capable of running a set of benchmarks"
+
+    def __init__(self, root: str, executor: CommandExecutor) -> None:
+        self.root = root
+        self.executor = executor
+
+    @abstractmethod
+    def run_benchmark_groups(self, benchmark_groups: List[BenchmarkGroup]) -> None:
+        "Run a list of benchmark groups"
+
+
+class ConbenchBenchmarkGroupsRunner(BenchmarkGroupsRunner):
+    "Runner class for conbench runner"
+
+    def run_benchmark_groups(self, benchmark_groups: List[BenchmarkGroup]) -> None:
+        Run.print_env_vars()
+        for benchmark_group in benchmark_groups:
+            self.run_benchmark_group(benchmark_group)
+            if benchmark_group.failed and benchmark_group.retry_on_failure:
+                print(f"Retrying {benchmark_group.command}")
+                self.run_benchmark_group(benchmark_group)
+
+    def run_benchmark_group(self, benchmark_group: BenchmarkGroup) -> None:
+        benchmark_group.started_at = datetime.now()
+        benchmark_group.log_execution()
+        benchmark_group.start_memory_monitor()
+
+        return_code, stderr = self.executor.execute_command(
+            benchmark_group.command,
+            path=self.root,
+            exit_on_failure=False,
+            log_stdout=(
+                benchmark_group.lang != "Java"
+            ),  # Java benchmarks produce 12GB+ of output
+        )
+
+        benchmark_group.finished_at = datetime.now()
+        benchmark_group.return_code = return_code
+        benchmark_group.stderr = stderr
+        benchmark_group.log_execution()
+        benchmark_group.stop_memory_monitor()
+
+
+class ArrowbenchBenchmarkGroupsRunner(BenchmarkGroupsRunner):
+    "Runner class for arrowbench"
+
+    def run_benchmark_groups(self, benchmark_groups: List[BenchmarkGroup]) -> None:
+        Run.print_env_vars()
+        # NOTE: `bm.command` is the raw arrowbench name; `bm.name` is f"arrowbench/{bm.command}"
+        # to disambiguate from labs/benchmarks versions when filtering.
+        bm_names = [bm.command for bm in benchmark_groups]
+        r_command = f"""
+        arrowbench::install_pipx();
+        arrowbench::install_benchconnect();
```

Comment:
Added. I'm not sure if this is the best way to install benchconnect in this context—there are already a ton of environments in this repo, and we could just add it to the relevant one, if we could figure out which that is—but this works.

If we care about speed, it's probably worth switching from conda to mamba (or at least the mamba solver); we're currently repeatedly solving environments, and it's not fast.

Reply:
> If we care about speed, it's probably worth switching from conda to mamba (or at least the mamba solver); we're currently repeatedly solving environments, and it's not fast.

Would you mind making an issue in this repo for that? Agreed that's probably the right way (but we also probably shouldn't do this just yet unless it's super quick + easy)

Reply:
#103